modules/renter/download.go

```diff
@@ -1,182 +1,290 @@
+// chunkDownload tracks the progress of a chunk.
+type chunkDownload struct {
+	// Map of workers who have attempted downloads for this chunk, from the
+	// worker id to the index of the piece they are trying to download.
+	workerAttempts map[types.FileContractID]uint64
 
-	// fetch returns the data specified by piece metadata.
-	fetch(pieceData) ([]byte, error)
+	// List of completed pieces, mapping from the index of the piece to the
+	// data.
+	completedPieces map[uint64][]byte
 }
 
-// A hostFetcher fetches pieces from a host. It implements the fetcher
-// interface.
-type hostFetcher struct {
-	downloader contractor.Downloader
-	pieceMap   map[uint64][]pieceData
-	masterKey  crypto.TwofishKey
-}
+// A download is a file download that has been queued by the renter.
+type download struct {
+	atomicReceived   uint64
+	atomicChunkIndex uint64
+
+	startTime   time.Time
+	siapath     string
+	destination string
 
-// pieces returns the pieces stored on this host that are part of a given
-// chunk.
-func (hf *hostFetcher) pieces(chunk uint64) []pieceData {
-	return hf.pieceMap[chunk]
+	erasureCode modules.ErasureCoder
+	chunkSize   uint64
+	fileSize    uint64
+
+	downloadFinished chan error
 }
 
-// fetch downloads the piece specified by p.
-func (hf *hostFetcher) fetch(p pieceData) ([]byte, error) {
-	// request piece
-	data, err := hf.downloader.Sector(p.MerkleRoot)
-	if err != nil {
-		return nil, err
-	}
+// newDownload initializes and returns a download object.
+func (f *file) newDownload(destination string) *download {
+	d := &download{
+		atomicChunkIndex: 0,
+		atomicReceived:   0,
 
-	// generate decryption key
-	key := deriveKey(hf.masterKey, p.Chunk, p.Piece)
+		erasureCode: f.erasureCode,
+		chunkSize:   f.chunkSize(),
+		fileSize:    f.size,
 
-	// decrypt and return
-	return key.DecryptBytes(data)
+		startTime:   time.Now(),
+		siapath:     f.name,
+		destination: destination,
+
+		downloadFinished: make(chan error),
+	}
+	return d
 }
 
-// newHostFetcher creates a new hostFetcher.
-func newHostFetcher(d contractor.Downloader, pieces []pieceData, masterKey crypto.TwofishKey) *hostFetcher {
-	// make piece map
-	pieceMap := make(map[uint64][]pieceData)
-	for _, p := range pieces {
-		pieceMap[p.Chunk] = append(pieceMap[p.Chunk], p)
+// managedDownloadIteraiton downloads a chunk from the next available file.
+func (r *Renter) managedDownloadIteration() {
+	// Get the set of available workers.
+	availableWorkers := make([]*worker, 0)
+	id := r.mu.RLock()
+	for _, worker := range r.workerPool {
+		availableWorkers = append(availableWorkers, worker)
 	}
-	return &hostFetcher{
-		downloader: d,
-		pieceMap:   pieceMap,
-		masterKey:  masterKey,
+	r.mu.RUnlock(id)
+
+	// Grab the next file to be downloaded.
+	id = r.mu.RLock()
+	if len(r.downloadQueue) == 0 {
+		// TODO: Block until either there's another download, or until
+		// shutdown, or until enough time has passed that the workers should be
+		// reassembled?
```

Comment:
yeah, the way it's currently coded it will spin. I have not fully decided the best way to handle blocking here.